File: c/grains/grains.c
#include "grains.h"
#define CHESS_SQUARES 64

uint64_t square(uint8_t index)
{
   if (index < 1 || index > CHESS_SQUARES)
   {
      return 0;
   }
   return (uint64_t)1 << (index - 1);
}

uint64_t total(void)
{
   int count = 0;
   for (int i = 1; i <= CHESS_SQUARES; ++i)
   {
      count += square(i);
   }
   return count;
}

File: haskell/grains/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n
  | n >= 1 && n <= 64 = Just (2 ^ (n - 1))
  | otherwise = Nothing

total :: Integer
total = case sum <$> mapM square [1 .. 64] of
  Just n -> n
  Nothing -> error "Invalid input"
